### PR TITLE
[CR] C++20 migration

### PR DESCRIPTION
**Pull request by scarf005**
#### Summary

SUMMARY: Build "Migrate to C++20"

#### Purpose of change
- continuation of #1624

#### Describe the solution
- manually patched workflow files and errors

#### Testing
- for some cursed reason `sudo apt install g++-12` is converted to `sudo apt install clang++12` in workflow

### CODE REVIEW

File: src/vpart_position.h [hunk 1]
```diff
@@ -48,9 +48,13 @@ class vpart_position
             return vehicle_.get();
         }
         // TODO: remove this, add a vpart_reference class instead
+        // HACK: don't know how to handle this
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
         size_t part_index() const {
             return part_index_;
         }
+#pragma GCC diagnostic pop
```
Comment (scarf005) [inline src/vpart_position.h:57]: not sure how to handle this, need help

Comment (Coolthulhu) [reply src/vpart_position.h:57]: You could assign a default value (some huge number) to `part_index_`.
GCC is known to produce false positives for this warning.

Comment (andrei8l) [reply src/vpart_position.h:57]: See BrettDong's solution from CDDA https://github.com/CleverRaven/Cataclysm-DDA/pull/58210

**Comment by scarf005** [comment]
seems like apple clang does not support ranges library yet. will have to migrate build to llvm

**Pull request closed without merging**